cirq-core/cirq/ops/greedy_qubit_manager.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Iterable, List, Set, TYPE_CHECKING
+
+from cirq.ops import named_qubit, qid_util, qubit_manager
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class GreedyQubitManager(qubit_manager.QubitManager):
+    """Greedy allocator that maximizes/minimizes qubit reuse based on a configurable parameter.
+
+    GreedyQubitManager can be configured, using `maximize_reuse` flag, to work in one of two modes:
+    - Minimize qubit reuse (maximize_reuse=False): For a fixed width, this mode uses a FIFO (First
+             in First out) strategy s.t. next allocated qubit is one which was freed the earliest.
+    - Maximize qubit reuse (maximize_reuse=True): For a fixed width, this mode uses a LIFO (Last in
+            First out) strategy s.t. the next allocated qubit is one which was freed the latest.
+
+    If the requested qubits are more than the set of free qubits, the qubit manager automatically
+    resizes the size of the managed qubit pool and adds new free qubits, that have their last
+    freed time to be -infinity.
+
+    For borrowing qubits, the qubit manager simply delegates borrow requests to `self.qalloc`, thus
+    always allocating new clean qubits.
+    """
+
+    def __init__(self, prefix: str, *, size: int = 0, maximize_reuse: bool = False):
+        """Initializes `GreedyQubitManager`
+
+        Args:
+            prefix: The prefix to use for naming new clean ancillas allocated by the qubit manager.
+                    The i'th allocated qubit is of the type `cirq.NamedQubit(f'{prefix}_{i}')`.
+            size: The initial size of the pool of ancilla qubits managed by the qubit manager. The
+                    qubit manager can automatically resize itself when the allocation request
+                    exceeds the number of available qubits.
+            maximize_reuse: Flag to control a FIFO vs LIFO strategy, defaults to False (FIFO).
+        """
+        self._prefix = prefix
+        self._used_qubits: Set['cirq.Qid'] = set()
+        self._free_qubits: List['cirq.Qid'] = []
```

Comment:
let the record reflect that I tried to actually verify that this is just a move but you've changed a few things :p including quoting the types now

Reply:
Quoting the types was necessary because earlier, in Cirq-FT, we could just do `import cirq` but now we either need to import the submodules or put `import cirq` under `if TYPE_CHECKING` and quote the types. 

This should be the only change, I haven't made any logic changes as part of the move.